lib/timeout.rb

```diff
@@ -164,6 +165,7 @@ def self.ensure_timeout_thread_created
   # Timeout</tt> into your classes so they have a #timeout method, as well as
   # a module method, so you can call it directly as Timeout.timeout().
   def timeout(sec, klass = nil, message = nil, &block)   #:yield: +sec+
+    raise ArgumentError, "Timeout sec must be a non-negative number" if sec && !(sec.is_a?(Numeric) && sec >= 0)
     return yield(sec) if sec == nil or sec.zero?
```

Comment:
I think we should switch the order of this check.  I don't think we should explicitly check for non-numeric input.

```suggestion
    return yield(sec) if sec == nil or sec.zero?
    raise ArgumentError, "Timeout sec must be a non-negative number" if 0 > sec
```

Reply:
@jeremyevans 

`I don't think we should explicitly check for non-numeric input.`

```
a = "test input"
=> "test input"
irb(main):027:0> 0 > a
Traceback (most recent call last):
        3: from /usr/bin/irb:11:in `<main>'
        2: from (irb):27
        1: from (irb):27:in `>'
ArgumentError (comparison of Integer with String failed)
```
This is again the same problem of inconsistent and less descriptive error msg

Reply:
`I think we should switch the order of this check`

isn't early validation better?

Reply:
Most CRuby core and stdlib methods do not attempt to validate all types up front:

```ruby
[]["a"]
# => no implicit conversion of String into Integer (TypeError)

StringIO.new("a").pos = "1"
# => no implicit conversion of String into Integer (TypeError)
```

Doing so is bad Ruby style.  In Ruby, as long as the object responds to the appropriate methods, it should work.

In this particular case, the issue is behavior for negative timeouts, so that is the only thing we should be checking for.  Having the negative check after the check for 0/nil simplifies the check, and improves performance in the 0/nil case.